Mutual TLS support

jon-wei opened the pull request:
This PR adds support for mutual TLS (https://en.wikipedia.org/wiki/Mutual_authentication). 

This adds a new "requireClientAuth" property and truststore properties on the server-side TLS config. 

If "requireClientAuth"  is true, Druid will require clients to identify themselves by providing a TLS certificate.

The server-side truststore needs to be configured with the certificates needed to validate the client certificates.

On the client side for internal communications (`simple-client-sslcontext` extension), new keystore properties are added, specifying the client certificate that a Druid service will use when communicating with other Druid services.

gianm commented:
@jon-wei, thanks, I think it is really valuable especially as we add more features to the feature.

(I heard you like features so we put features in your features?)

gianm commented:
@jon-wei Are you done working on this patch? I wasn't sure since the last few commits are named "WIP"

jon-wei commented:
Not yet, I made the integration tests use mutual TLS, but there's a test for coordinator -> overlord proxying that's failing, haven't fixed that yet

gianm commented:
Got it, thanks.

jon-wei commented:
I updated the patch with the following:
- Added mutual TLS to the integration tests. All druid services will have both plaintext and TLS port enabled with client cert checking enabled
- Adjusted the timers in the integration tests: I shortened the retry periods and increased the number of retries
- Fixed some timing issues in certain integration tests (like feeding events to a worker thread before it's ready)
- Fixes coordinator -> overlord proxy requests when TLS is enabled (the OverlordProxyServlet was not using a http client with the proper SSLContext)

fjy commented:
👍

gianm commented on docs/content/operations/tls-support.md:
```diff
@@ -31,6 +31,16 @@ values for the below mentioned configs among others provided by Java implementat
+|Property|Description|Default|Required|
+|--------|-----------|-------|--------|
+|`druid.server.https.requireClientAuth`|If set to true, clients must identify themselves by providing a TLS certificate.  If `requireClientAuth` is false, the rest of the options in this table are ignored.|false|no|
+|`druid.server.https.trustStoreType`|The type of the trust store containing certificates used to validate client certificates.|`java.security.KeyStore.getDefaultType()`|no|
```
This won't be used for anything other than verifying client certificates?

jon-wei replied:
yes, the truststore on the server configuration side is only used to check the client certificates

gianm commented on docs/content/operations/tls-support.md:
```diff
@@ -31,6 +31,16 @@ values for the below mentioned configs among others provided by Java implementat
+
+|Property|Description|Default|Required|
+|--------|-----------|-------|--------|
+|`druid.server.https.requireClientAuth`|If set to true, clients must identify themselves by providing a TLS certificate.  If `requireClientAuth` is false, the rest of the options in this table are ignored.|false|no|
```
How about `requireClientCertificate`?

jon-wei replied:
Changed to `requireClientCertificate`

jon-wei commented:
> Different private keys and certs for the client and server. (Can be the same root CA, and same CN/altNames, but generate two different keys, and two different keyStores.)

The services and clients have different keys, I'll update the patch to rename the client files so that this is more apparent

gianm commented:
@jon-wei is this still a WIP or is it ready for re-review?

jon-wei commented:
@gianm Still WIP, adjusting the networking in the integration tests for hostname verification, also looking into cert revocation handling

jon-wei commented:
Added the following changes:
- hostname validation and static CRL options for client certificate authentication
- Fix for starting mysql on when not using docker-machine with docker overlay2 fs
- Negative tests for:
  - No client cert
  - Client cert signed by untrusted root
  - Client cert signed by an intermediate cert that is not a CA
  - Client cert with wrong hostname
  - Expired client cert
  - Client cert has been revoked in static CRL file
- Positive test for:
  - 3-part cert chain (root->intermediate->client cert)
- Added two more services to the integration tests:
  - Router with permissive TLS client auth settings (no hostname verification or revocation)
  - Router with TLS client auth turned off

jon-wei commented:
@nishantmonu51 @himanshug @gianm 

My apologies for the large number of post-review changes, this is ready for review again

jon-wei commented:
Marking WIP again, some networking changes introduced broke Kafka functionality within the tests

gianm commented on server/src/main/java/org/apache/druid/server/StatusResource.java:
```diff
@@ -47,6 +50,7 @@
 @Path("/status")
 public class StatusResource
 {
+  private static Logger log = new Logger(StatusResource.class);
```
Doesn't look necessary.

jon-wei replied:
Removed

gianm commented on server/src/main/java/org/apache/druid/server/security/TLSUtils.java:
```diff
@@ -39,28 +44,77 @@ public static SSLContext createSSLContext(
+      String keyStoreAlgorithm,
+      String certAlias,
+      PasswordProvider keyStorePasswordProvider,
+      PasswordProvider keyManagerFactoryPasswordProvider
```
This method's arg list is getting super long. It would be nicer, although not required, to use a builder instead.

jon-wei replied:
Added a builder and replaced constructor calls with builder use

gianm commented on docs/content/operations/tls-support.md:
```diff
@@ -31,7 +31,19 @@ values for the below mentioned configs among others provided by Java implementat
+|`druid.server.https.trustStorePath`|The file path or URL of the trust store containing certificates used to validate client certificates. Not needed if `requireClientCertificate` is false.|none|yes, only if `requireClientCertificate` is true|
+|`druid.server.https.trustStoreAlgorithm`|Algorithm to be used by TrustManager to validate client certificate chains. Not needed if `requireClientCertificate` is false.|`javax.net.ssl.TrustManagerFactory.getDefaultAlgorithm()`|no|
+|`druid.server.https.trustStorePassword`|The [Password Provider](../../operations/password-provider.html) or String password for the Trust Store.  Not needed if `requireClientCertificate` is false.|none|no|
+|`druid.server.https.validateHostnames`|If set to true, perform hostname validation on client certificates.  Not used if `requireClientCertificate` is false.|false|no|
```
What does "hostname validation" mean in this context? If it means checking CN/altNames, then it should be true by default, since otherwise it's a trap waiting to be sprung on people.

(Once you turn on `druid.server.https.requireClientCertificate`, we should be secure as possible by default, and you should have to explicitly set options to be less secure.)

jon-wei replied:
Changed default to true and added more details to the documentation

jon-wei commented on server/src/main/java/org/apache/druid/server/initialization/jetty/JettyServerModule.java:
```diff
@@ -256,6 +259,32 @@ static Server makeAndInitializeServer(
+        sslContextFactory.setNeedClientAuth(tlsServerConfig.isRequireClientCertificate());
+        if (tlsServerConfig.isRequireClientCertificate()) {
+          if (tlsServerConfig.getCrlPath() != null) {
+            sslContextFactory.setValidatePeerCerts(true);
```
Added a comment here

jon-wei commented on server/src/main/java/org/apache/druid/server/initialization/jetty/JettyServerModule.java:
```diff
@@ -256,6 +259,32 @@ static Server makeAndInitializeServer(
+          if (tlsServerConfig.getTrustStorePath() != null) {
+            sslContextFactory.setTrustStorePath(tlsServerConfig.getTrustStorePath());
+            sslContextFactory.setTrustStoreType(
+                tlsServerConfig.getTrustStoreType() == null ?
```
Reformatted these areas

jon-wei commented on server/src/main/java/org/apache/druid/server/security/TLSUtils.java:
```diff
@@ -39,28 +44,77 @@ public static SSLContext createSSLContext(
       String trustStoreType,
```
Marked the nullable params @Nullable

jon-wei commented on server/src/main/java/org/apache/druid/server/security/TLSUtils.java:
```diff
@@ -39,28 +44,77 @@ public static SSLContext createSSLContext(
-      PasswordProvider trustStorePasswordProvider
+      PasswordProvider trustStorePasswordProvider,
+      String keyStoreType,
+      String keyStorePath,
```
Marked the nullable params @Nullable

jon-wei commented:
Addressed comments, also adjusted `pom.xml` to exclude the TLS certs/keys generated by the integration tests from Rat license chcking
